src/porepy/numerics/ad/_ad_utils.py

```diff
@@ -447,6 +447,51 @@ def get_solution_values(
     return value
 
 
+def shift_solution_values(
+    name: str,
+    data: dict,
+    location: Any,
+) -> None:
+    """Function to shift numerical values stored in the data dictionary.
+
+    The shift is implemented s.t. values at index ``i`` are copied to index ``i + 1``,
+    based on the number of stored values.
+
+    Note:
+        The data stored must have support for ``.copy()`` in order to avoid faulty
+        referencing (e.g., numpy arrays or sparse matrices).
+
+    Parameters:
+        name: Key in ``data`` for which quantity the shift should be performed.
+        data: A grid data dictionary.
+        location: Either :data:`~porepy.utils.common_constants.TIME_STEP_SOLUTIONS`
+            or :data:`~porepy.utils.common_constants.ITERATE_SOLUTIONS`.
+
+    Raises:
+        ValueError: If unsupported ``location`` is passed.
+
+    """
+    if location not in data:
+        data[location] = {}
+    if name not in data[location]:
+        # TODO should we return here instead? Nothing to be shifted, and
```

Comment:
I think I agree with not doing anything here. Such creation of dictionary items by stealth can be very confusing under debugging.

Reply:
Changed to return if no location or name in data dictionary found.